### PR TITLE
Vnode proxy stuck in overload state

### PR DESCRIPTION
**Pull request by jonmeredith**
Fix for https://github.com/basho/riak_core/issues/760 ([RIAK-1914](https://bashoeng.atlassian.net/browse/RIAK-1914)).

Previously, if a heavily loaded vnode received more than `check_threshold` msgs after it hit the direct mailbox check, the ping/pong would leave it in overload state.  In cases where the load was removed (e.g. primary recovered) the vnode proxy would stay in overload state until enough messages were received to trigger pinging the vnode or rechecking the mailbox.

This PR does a couple of things
1) Stops counting messages skipped due to overload in the mailbox estimate.
2) Changes the ping/pong mechanism to use a reference so we can be resubmit safely.
3) After doing a direct check, immediately trigger a ping/pong so that once the mailbox is cleared the proxy will be immediately notified.
4) Adds an `overloaded/1` function to make testing easier.

There is corresponding riak_test on branch bugfix/jdm/vnode-proxy-stuck-in-overload-v2

Customer ticket zd://11440

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
You can't retry until you :+1: a commit.

**Comment by cmeiklejohn** [comment]
:+1: c322af3

File: src/riak_core_vnode_proxy.erl [hunk 1]
```diff
@@ -317,6 +326,10 @@ fake_loop() ->
         {get_count, Pid} ->
             Pid ! {count, erlang:get(count)},
             fake_loop();
+        %% Original tests do not expect replies
```
Comment (cmeiklejohn) [inline src/riak_core_vnode_proxy.erl]: Remove commented code.

File: src/riak_core_vnode_proxy.erl [hunk 2]
```diff
@@ -399,15 +421,16 @@ overload_test_() ->
                fun() ->
                        VnodePid ! slow,
                        [ProxyPid ! hello || _ <- lists:seq(1, 50000)],
-                       %% synchronize on the mailbox
+                       %% synchronize on the mailbox - no-op that hits msg catchall
                        Reply = gen:call(ProxyPid, '$vnode_proxy_call', sync, infinity),
                        ?assertEqual({ok, ok}, Reply),
```
Comment (cmeiklejohn) [inline src/riak_core_vnode_proxy.erl:427]: Love the `{ok, ok}`, _snorts_.

File: src/riak_core_vnode_proxy.erl (hunk 1 above)
Comment (jonmeredith) [inline src/riak_core_vnode_proxy.erl]: Actually added this for future souls on purpose as I couldn't work out where my replies were.  I'll improve the comment.

File: src/riak_core_vnode_proxy.erl (hunk 2 above)
Comment (jonmeredith) [inline src/riak_core_vnode_proxy.erl:427]: It's high strung.

**Comment by jonmeredith** [comment]
@cmeiklejohn updated comments per review.  Would you mind +1ing the new SHA?

**Comment by jonmeredith** [comment]
Stupid racey test

```
**error:{assertEqual_failed,[{module,riak_core_vnode_proxy},
                     {line,397},
                     {expression,"Count"},
                     {expected,20005},
                     {value,19569}]}
```

Put the limit back so it cannot overload.  EQC test coverage is much better anyway.

**Comment by jonmeredith** [comment]
@cmeiklejohn once more, with feeling please :)

**Comment by cmeiklejohn** [comment]
:+1: 63ff909

**Comment by jonmeredith** [comment]
@borshop merge

**Pull request merged**